crates/matrix-sdk-ui/src/room_list_service/mod.rs

```diff
@@ -435,7 +437,7 @@ impl RoomListService {
 
         let room = Room::new(&self.client, room_id, &self.sliding_sync)?;
 
-        self.rooms.write().await.push(room.clone());
+        self.rooms.write().unwrap().push(room.clone());
```

Comment:
Preexisting: this has one subtle race conditions, when two callers call around the same time and there's no cache hit in `self.rooms`. After taking the write-lock, you need to check for the room being there again, probably before calling `Room::new()`.

An alternative is also to take the lock as writable from the beginning, and keep it for the entire function's body.

Reply:
You're right! Let's change the `RwLock` for a `Mutex` then. It's now cheap to create a `Room`, which wasn't the case before because of the `.await` point.